optuna/integration/cma.py

```diff
@@ -0,0 +1,444 @@
+
+        n_startup_trials:
+            The independent sampling is used instead of the CMA-ES algorithm until the given number
+            of trials finish in the same study.
+
+        independent_sampler:
+            A :class:`~optuna.samplers.BaseSampler` instance that is used for independent
+            sampling. The parameters not contained in the relative search space are sampled
+            by this sampler.
+            The search space for :class:`~optuna.integration.CmaEsSampler` is determined by
+            :func:`~optuna.samplers.product_search_space()`.
+
+            If :obj:`None` is specified, :class:`~optuna.samplers.RandomSampler` is used
+            as the default.
+
+            .. seealso::
+                :class:`optuna.samplers` module provides built-in independent samplers
+                such as :class:`~optuna.samplers.RandomSampler` and
+                :class:`~optuna.samplers.TPESampler`.
+
+        warn_independent_sampling:
+            If this is :obj:`True`, a warning message is emitted when
+            the value of a parameter is sampled by using an independent sampler.
+
+            Note that the parameters of the first trial in a study are always sampled
+            via an independent sampler, so no warning messages are emitted in this case.
+    """
+
+    def __init__(
+            self,
+            x0=None,  # type: Optional[Dict[str, Any]]
+            sigma0=None,  # type: Optional[float]
+            cma_stds=None,  # type: Optional[Dict[str, float]]
+            seed=None,  # type: Optional[int]
+            cma_opts=None,  # type: Optional[Dict[str, Any]]
+            n_startup_trials=1,  # type: int
+            independent_sampler=None,  # type: Optional[BaseSampler]
+            warn_independent_sampling=True,  # type: bool
+    ):
+        # type: (...) -> None
+
+        _check_cma_availability()
+
+        self._x0 = x0
+        self._sigma0 = sigma0
+        self._cma_stds = cma_stds
+        if seed is None:
+            seed = random.randint(1, 2**32)
+        self._cma_opts = cma_opts or {}
+        self._cma_opts['seed'] = seed
+        self._cma_opts.setdefault('verbose', -2)
+        self._n_startup_trials = n_startup_trials
+        self._independent_sampler = independent_sampler or optuna.samplers.RandomSampler(seed=seed)
+        self._warn_independent_sampling = warn_independent_sampling
+        self._logger = optuna.logging.get_logger(__name__)
+
+    def infer_relative_search_space(self, study, trial):
+        # type: (InTrialStudy, FrozenTrial) -> Dict[str, BaseDistribution]
+
+        search_space = {}
+        for name, distribution in optuna.samplers.product_search_space(study).items():
+            if distribution.single():
+                # `cma` cannot handle distributions that contain just a single value,
+                # so we skip this distribution.
+                #
+                # Note that `Trial` takes care of this distribution during suggestion.
```

Comment:
I think `takes care of` too redundant to grasp what a `Trial` actually does.

Reply:
Thank you for pointing it out. I updated this comment as follows since the current `Trial` does not seem to have code to handle distributions with a single value. If I misunderstood the behavior or `Trial`, please let me know.

```
# `cma` cannot handle distributions that contain just a single value, so that
# the parameter values for such distributions are sampled by `sample_independent`.
```

Reply:
As a result of offline discussion with @sile, I fixed the comment as follows:
```
# `cma` cannot handle distributions that contain just a single value, so we skip
# them. Note that the parameter values for such distributions are sampled in
# `Trial`.
```

The `Trial` actually has code to handle distributions with a single value at
https://github.com/pfnet/optuna/blob/master/optuna/trial.py#L178, https://github.com/pfnet/optuna/blob/master/optuna/trial.py#L218 and so on.